rfc/rfc-79/rfc-79.md

```diff
@@ -0,0 +1,138 @@
+We heard from uber, that this stray executor could sometimes be running even after spark application is completed (very one off case)
+
+#### Case2:
+Dag retrigger while writing to MDT(spark rdd cache invalidation). Issue exacerbated with RLI.
+
+1 -> 2 … 5(writes) -> 6(marker based reconciliation) ->
+After marker based reconciliation is complete, while trying to write to MDT, say the dag is retriggered due to spark rdd cache invalidation. So, write handles could be re-executed, but the reconciliation may not get executed again. So, this could lead to duplicate data files on storage. Infact our commit metadata could be tracking one set of files, but MDT writes could be tracking a diff set of files.
+
+
+## Design
+
+We are proposing two additional markers to assist with detecting and thwarting spurious data files in Hudi on any cost.
+### 1. Completion Marker:
+   The marker mechanism discussed above strictly speaking is a begin marker. So, lets add a completion maker at the end of writes (write handle) for each data file produced to mark the completion of the write for a given task in addition to the begin marker. Completion marker will only track the fileId and commit time(write token should be ignored) and the writeStatus gets serialized as content. This might be in contrast to the begin markers where the marker will track the actual data file name.
+
+Write handle changes:
+In the  constructor
+1. If completion marker is already present for the given fileID and commit time, read the CM(completion marker), to deser write status and return right away. (short circuit).
+2. If completion marker is not present, go ahead w/ writing the data file(s).
+3. In the end just before returning the writeStatus back to driver:
+   again check for CM.
+   if present, delete the newly created file and read the CM, to deser write status and return the same WriteStatus.
+   if not, go ahead, create CM and return WriteStatus. or spark could have aborted the task by then in which case there won't be any data file only.
+
+Most of spark task retries, executor losses should be taken care by the marker based reconciliation even w/o completion marker fix. Above design should help w/ stray executor issues and spark speculative execution.
```

Comment:
I believe someone had suggested the idea of using the timeline server, since then hypothetically you could have a setup where each file-writing spark task has to get an ACK from the timeline server before starting any file writes, and informs the timeline server when it is done (like some sort of simplified lease). And this way you can prevent spark tasks from creating files after the finalize-write check. Is there a reason why we are not pursuing this route? I guess there is the issue that you have to handle cases where a single spark task failure can cause the entire stage/finalize write check to fail (due to telling the Timeline server that it has started working on a file, but getting stuck after).